### PR TITLE
Improve `Skeleton3D::find_bone()` performance

### LINKED ISSUE #75915: Skeleton3D find_bone() is too slow
### Godot version

c3ed7af

### System information

Windows 10

### Issue description

The Skeleton.find_bone() function is too slow with complex skeletons that have many bones.

While this was always the issue with find_bone() even far back as in Godot 3 the issue with its slowness is now amplified due to other more recent changes to skeletons and the asset import.

E.g. in Godot 4 the asset importer adds extra bones for each bone attachment. You end up with a lot more bones than usual to search. If you have 100 meshes all parented to the same bone in e.g. Blender you end up with 100 extra bones in Godot.

Godot 3 did not add extra bones for each bone attachment but this was apparently changed to solve other bugs with index shuffling like https://github.com/godotengine/godot/issues/64512 and https://github.com/godotengine/godot/issues/61641.

The old workaround of using the bone index instead of the name to do e.g. runtime skeleton mods does not work well now that the bone index are shuffled still with every changed attachment that adds new bones. The current situation requires to use a custom cache for bone name to bone index instead of using the slow Skeleton find_bone() function.

The solution could possibly be to use an internal HashMap for bone_name_to_bone_index instead of the current find_bone() loop that goes over the entire bone array everytime.

### Steps to reproduce

```gdscript
func _ready() -> void:
	var skeleton : Skeleton3D = Skeleton3D.new()
	var bone_names : Array = []
	for i in 500:
		var bone_name : String = "bone_%s" % str(i)
		bone_names.push_back(bone_name)
		skeleton.add_bone(bone_names[i])
	
	var bone_map : Dictionary = {}
	for bone_name in bone_names:
		var bone_index : int = skeleton.find_bone(bone_name)
		bone_map[bone_name] = bone_index
	
	var time : int = 0
	var time_total : int = 0
	
	time = Time.get_ticks_usec()
	for bone_name in bone_names:
		var bone_index : int = skeleton.find_bone(bone_name)
	time_total = Time.get_ticks_usec() - time
	print("Time find_bone(): %s" % time_total)
	
	time = Time.get_ticks_usec()
	for bone_name in bone_names:
		var bone_index : int = bone_map[bone_name]
	time_total = Time.get_ticks_usec() - time
	print("Time Dictionary: %s" % time_total)

	# E.g. result:
	# Time find_bone(): 5907
	# Time Dictionary: 201
```

### Minimal reproduction project

see steps

### PR DESCRIPTION
**Pull request by 44zb**
Indexing Skeleon3D bones using HashMap

results of the script from #75915 before/after this fix
before
Time find_bone(): 1055
Time Dictionary: 59

After	
Time find_bone(): 65
Time Dictionary: 57

Fix #75915

### CODE REVIEW

File: scene/3d/skeleton_3d.cpp [hunk 1]
```diff
@@ -400,31 +397,31 @@ void Skeleton3D::add_bone(const String &p_name) {
...
 
 int Skeleton3D::find_bone(const String &p_name) const {
-	for (int i = 0; i < bones.size(); i++) {
-		if (bones[i].name == p_name) {
-			return i;
-		}
+	HashMap<String, int>::ConstIterator index = name_to_bone_index.find(p_name);
+	if (index == name_to_bone_index.end()) {
+		return -1;
 	}
```
Comment (kleonc) [inline scene/3d/skeleton_3d.cpp:404]:
```suggestion
	int* bone_index_ptr = name_to_bone_index.getptr(p_name);
	return bone_index_ptr != nullptr ? *bone_index_ptr : -1;
```
```

**Review by kleonc — COMMENTED** [review]
@44zb LGTM besides the PR/commit title. No need for these to link to the issue being fixed (#75915)(it could be done in the PR/commit description). Also more human-readable titles are preferred. So I suggest changing both PR title and commit title (first line of the commit message) to e.g. *"Improve Skeleton3D::find_bone() performance"*.

Also seems like there's some formatting issue, should be addressed too (see [the failing test](https://github.com/godotengine/godot/actions/runs/5058619134/jobs/9079038408?pr=77307))(Yeah, I suggested it like that, oops! :upside_down_face:).

**Comment by kleonc** [comment]
@44zb Oh, you actually still haven't changed [the commit](https://github.com/godotengine/godot/pull/77307/commits/8cccf7f1ecf81d1ced3f172c0ad4e5a9d40f0321) message, it's still *"Skeleton3D::find_bone() performance https://github.com/godotengine/godot/issues/75915"*. You can change it when rebasing by using `-r`/`--reword` for the given commit.

**Comment by 44zb** [comment]
@kleonc yes, sorry missed the commit part. But now it's all good :)

**Review by kleonc — APPROVED** [review]

**Comment by fire** [comment]
As long as this works for all the input cases and the performance is faster I'm fine with it.

I don't think this portion of the code requires special memory allocation strategies like not allocating.

**Review by clayjohn — APPROVED** [review]
Looks great!

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Comment by YuriSizov** [comment]
Cherry-picked for 4.0.4 with https://github.com/godotengine/godot/pull/77874.

**Pull request merged**